job.c

```diff
@@ -34,24 +34,32 @@ store_job(job j)
     all_jobs_used++;
 
     /* accept a load factor of 4 */
-    if (all_jobs_used > (all_jobs_cap << 2)) rehash();
+    if (all_jobs_used > (all_jobs_cap << 2)) rehash(1);
 }
 
 static void
-rehash()
+rehash(int is_upscaling)
 {
     job *old = all_jobs;
     size_t old_cap = all_jobs_cap, old_used = all_jobs_used, i;
+    int old_prime = cur_prime;
 
-    if (cur_prime >= NUM_PRIMES) return;
-    if (hash_table_was_oom) return;
+    if (is_upscaling) {
```

Comment:
Please consider defining an index delta and using that to find the next prime:

```
int d = is_upscaling ? 1 : -1;

if (cur_prime + d >= NUM_PRIMES) return;
if (cur_prime + d < 0) return;
if (is_upscaling && hash_table_was_oom) return;

cur_prime += d;
```

This will make the code a little easier to follow

(I also sneakily fixed an off-by-one error here that was present in the original code, in the comparison against NUM_PRIMES.)